### DISCUSSION TITLE
how to bind trace and log

### DISCUSSION THREAD

**Discussion by Jedore** [apache/skywalking#9371] [Q&A] [answered]
SkyWalking version: 9.0.0
Python Agent: 0.8.0
FastAPI: 0.72.0
Now i have a api trace `52c1ad20098211ed99cc70b5e8357f07`
![image](https://user-images.githubusercontent.com/17944025/180377114-4952d98a-3269-42b8-8708-2f012c0e06cc.png)
and i have a log `52c1ad20098211ed99cc70b5e8357f07`
![image](https://user-images.githubusercontent.com/17944025/180377271-b253b62b-0501-46e4-a750-af3842374109.png)
The problem is i can't view logs in trace ui,
![image](https://user-images.githubusercontent.com/17944025/180377570-be5c6122-d3d0-4730-a76b-395eda4902a0.png)
after clicked button `view logs`, i saw nothing
![image](https://user-images.githubusercontent.com/17944025/180377697-7b2bf92f-aa51-4b78-a139-da46d99c8704.png)
Does skywalking support this and how can i fix it ? Thanks!

**Comment by wu-sheng** [comment]
Which agent do you use? I think generally it should support collecting logs. Or you have to print trace ID, span ID, segment ID in the logs, and use LAL to split them as structured log in the OAP.

**Reply by Jedore** [reply]
> Which agent do you use? I think generally it should support collecting logs. Or you have to print trace ID, span ID, segment ID in the logs, and use LAL to split them as structured log in the OAP.

Python Agent, in FastAPI. 2 steps:
1. config
```
config.init(collector_address='192.168.90.14:11800', service_name='test',
            log_reporter_active=True)  # defaults to grpc protocol
agent.start()
```
2. logging
```
@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    logger.error('item id')
    return {"item_id": item_id, "q": q}
```
No more operations.  Logs with trace id has been collected into oap, just not binded with traces

**Reply by wu-sheng** [reply]
Read docs, if python agent doesn't declare it supports, then it is not.

**Reply by Jedore** [reply]
In fact, i read foundations docs and did not find info about button `view logs` or binding trace and log. Of course, python agent doc only says collecting log, no more about how to display. I'm confused.

**Reply by Jedore** [reply]
I can search logs in log ui by trace id, But can't check log through trace ui `view log` button.

**Reply by wu-sheng** [reply]
@kezhenxu94 @Superskyyy Does Python agent support this?

**Reply by Superskyyy** [reply]
> @kezhenxu94 @Superskyyy Does Python agent support this?

@Jedore Yes Python agent does support log trace correlation. There are multiple traces, are you seeing normal(same) number of trace and logs as you access the endpoint? I can't tell from the screenshots.

**Reply by Superskyyy** [reply]
> I can search logs in log ui by trace id, But can't check log through trace ui `view log` button.

Btw the trace you clicked on isn't the same as the log traceID, can you find the one correct trace? You clicked the wrong one.

**Reply by Jedore** [reply]
> > I can search logs in log ui by trace id, But can't check log through trace ui `view log` button.
> 
> Btw the trace you clicked on isn't the same as the log traceID, can you find the one correct trace? You clicked the wrong one.

@Superskyyy Sorry for wrong screeshots. I am not sure understanding your meaning.
I get new trace and log:
![image](https://user-images.githubusercontent.com/17944025/180675018-81a14320-3b6d-419a-8fa5-ab414624a74b.png)
![image](https://user-images.githubusercontent.com/17944025/180675113-c99e4fc8-4bcf-45ff-9899-91acfe4b029f.png)
I still can't check log through trace ui `view logs` button.

**Reply by Superskyyy** [reply]
> I still can't check log through trace ui `view logs` button. 
> 
> 

Can you provide a minimally reproducible code example? That way it's easier to see if there's a bug.

**Reply by Jedore** [reply]
> Can you provide a minimally reproducible code example? That way it's easier to see if there's a bug.

OK. Skywalking: `9.0.0` ; FastAPI: `0.72.0`; Python Agent: `0.8.0`
```python
from typing import Union
import logging

from fastapi import FastAPI
import uvicorn

app = FastAPI()

from skywalking import agent, config

config.init(collector_address='192.168.90.14:11800', service_name='test',
            log_reporter_active=True)  # defaults to grpc protocol
agent.start()

logger = logging.getLogger()


@app.get("/")
def read_root():
    logger.error('hello word')
    return {"Hello": "World"}


@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    logger.error('item id')
    return {"item_id": item_id, "q": q}


if __name__ == '__main__':
    uvicorn.run("maintest:app", host='0.0.0.0', port=8000, reload=False)
```

**Reply by Superskyyy** [reply]
Thank you, I will take a look

**Comment by Jedore** [comment, marked as answer]
9.3.0 has no this problem.

**Comment by jampsen** [comment]
This problem still in 9.6.0 booster ui 。 When I click on the trace ID in the log page to  the Trace interface, the data is not displayed. I need to manually switch the service to the corresponding one to display the data correctly. 
When Trace page jump to log page is the same problem.
This issue does not exist in the UI of version 8.X.
![log-1](https://github.com/apache/skywalking/assets/35960370/727bd890-11b8-4d7d-b633-e14b377ca3d4)
![trace-1](https://github.com/apache/skywalking/assets/35960370/016d2ee5-596f-4397-a889-ad4e2a65b80f)
![trace-2](https://github.com/apache/skywalking/assets/35960370/0d792c9e-8ae8-475e-a9cd-d1aa968acedb)

**Reply by wu-sheng** [reply]
@Fine0830 Could you take a look at this? I feel that, when trace ID is applied, the service should be empty. In all other manual query cases, it is required.

**Reply by Fine0830** [reply]
From screenshots,  I think the service is wrong when the view dump from Log to Trace widget by clicking traceID. It is a bug, let me fix it later.

**Reply by wu-sheng** [reply]
Fixed by https://github.com/apache/skywalking-booster-ui/pull/343
